runtime/lua/vim/lsp/_watchfiles.lua

```diff
@@ -0,0 +1,486 @@
+      end, 100)
+    end
+  end
+end
+
+-- Never create watchers for directories matching these patterns. Similar
+-- to vscode's files.watcherExclude setting.
+local excludes = {
+  parse('**/.git/objects/**'),
+  parse('**/.git/subtree-cache/**'),
+  parse('**/node_modules/*/**'),
+  parse('**/.hg/store/**'),
+}
+
+local fsevent_ensure_recursive
+
+---@private
+--- Initializes a libuv fs_event, persistent when underlying inodes change.
+---
+---@param path string The path to watch.
+---@return uv.fs_event The started libuv fs_event handle.
+local function start_watch(path)
+  local fsevent, fserr = uv.new_fs_event()
+  assert(not fserr, fserr)
+  fsevent:start(path, { recursive = recursive_watch }, function(err, filename, events)
+    assert(not err, err)
+
+    local fullpath = filepath_join(path, filename)
+
+    local change_type = events.change and protocol.FileChangeType.Changed or 0
+    if events.rename then
+      local _, staterr, staterrname = uv.fs_stat(fullpath)
+      if staterrname == 'ENOENT' then
+        change_type = protocol.FileChangeType.Deleted
+      else
+        assert(not staterr, staterr)
+        change_type = protocol.FileChangeType.Created
+      end
+    end
+
+    for _, regs in pairs(watched_paths[path].callbacks) do
+      for _, reg in pairs(regs) do
+        reg.callback(fullpath, change_type)
+      end
+    end
+
+    local stat, staterr, staterrname = uv.fs_stat(path)
+    if staterrname == 'ENOENT' then
+      local _, stoperr = fsevent:stop()
+      assert(not stoperr, stoperr)
+      fsevent:close()
+      watched_paths[path] = nil
+      return
+    end
+    assert(not staterr, staterr)
+    if stat.ino ~= watched_paths[path].inode then
+      watched_paths[path].inode = stat.ino
+      local _, stoperr = fsevent:stop()
+      assert(not stoperr, stoperr)
+      fsevent:close()
+      watched_paths[path].fsevent = start_watch(path)
+    end
+
+    if not recursive_watch and stat.type == 'directory' then
+      for client_id, regs in pairs(registrations) do
+        for reg_id, watchers in pairs(regs) do
+          for _, watcher in ipairs(watchers) do
+            if vim.startswith(path .. path_sep, watcher.base_dir .. path_sep) then
+              fsevent_ensure_recursive(
+                path,
+                watcher.pattern,
+                watcher.kind,
+                client_id,
+                reg_id,
+                stat.type
+              )
+            end
+          end
+        end
+      end
+    end
+  end)
+  return fsevent
+end
+
+---@private
+--- Creates libuv fs_events handles.
+---
+---@param path string Absolute path to watch.
+---@param pattern string|table The LSP glob pattern (raw or parsed) to match against.
+---@param kind number The LSP WatchKind value.
+---@param client_id number The LSP client's ID.
+---@param reg_id string The ID used to register the request.
+function fsevent_ensure_recursive(path, pattern, kind, client_id, reg_id, filetype)
```

Comment:
this is a ton of logic that is not lsp-specific. But it looks like you have put some thought into it, which could potentially be exposed as a generic interface for things like [better autoread](https://github.com/neovim/neovim/issues/1380#issuecomment-546667155).

Reply:
You're absolutely right that most of this isn't inherently LSP. I'll aim to use the general purpose watcher in #20801 as a basis for decoupling these.